src/librustc_mir/hair/pattern/_match.rs

```diff
@@ -447,43 +614,649 @@ impl<'tcx> Constructor<'tcx> {
+                    VarLenSlice(prefix, suffix) => prefix + suffix <= self_len,
+                    _ => false,
+                };
+                if used_ctors.iter().any(overlaps) { smallvec![] } else { smallvec![self] }
+            }
+            VarLenSlice(self_prefix, self_suffix) => {
+                // Assume we have the following match:
+                // ```
+                // match slice {
+                //     [0] => {}
+                //     [_, _, _] => {}
+                //     [1, 2, 3, 4, 5, 6, ..] => {}
+                //     [_, _, _, _, _, _, _] => {}
+                //     [0, ..] => {}
+                // }
+                // ```
+                // We want to know which constructors are matched by the last pattern, but are not
+                // matched by the first four ones. Since we only speak of constructors here, we
+                // only care about the length of the slices and not the particular subpatterns.
+                // For that, we first notice that because of the third pattern, all constructors of
+                // lengths 6 or more are covered already. `max_len` will be `Some(6)`.
+                // Then we'll look at fixed-length constructors to see which are missing. The
+                // returned list of constructors will be those of lengths in 1..6 that are not
+                // present in the match. Lengths 1, 3 and 7 are matched already, so we get
+                // `[FixedLenSlice(2), FixedLenSlice(4), FixedLenSlice(5)]`.
+                // If we had removed the third pattern, we would have instead returned
+                // `[FixedLenSlice(2), FixedLenSlice(4), FixedLenSlice(5), FixedLenSlice(6),
+                // VarLenSlice(8, 0)]`.
+
+                // Initially we cover all slice lengths starting from self_len.
+                let self_len = self_prefix + self_suffix;
+
+                // If there is a VarLenSlice(n) in used_ctors, then we have to discard
+                // all lengths >= n. So we pick the smallest such n.
+                let max_len: Option<_> = used_ctors
+                    .iter()
+                    .filter_map(|c: &Constructor<'tcx>| match c {
+                        VarLenSlice(prefix, suffix) => Some(prefix + suffix),
+                        _ => None,
+                    })
+                    .min();
+
+                // The remaining range of lengths is now either `self_len..`
+                // or `self_len..max_len`. We then remove from that range all the
+                // individual FixedLenSlice lengths in used_ctors.
+
+                // If max_len <= self_len there are no lengths remaining.
+                if let Some(max_len) = max_len {
+                    if max_len <= self_len {
+                        return smallvec![];
+                    }
+                }
+
+                // Extract fixed-size lengths
+                let used_fixed_lengths: FxHashSet<u64> = used_ctors
```

Comment:
Now that I noticed that constant patterns are a Thing, this logic looks a bit suspect.

Reply:
I'll just cc #65413, this isn't the only case that has problems.

Reply:
(add a comment here)